### ISSUE TITLE
Drone terminates with Segmentation fault

### ISSUE THREAD

**Issue by b-ehlers** [pstavirs/ostinato#280]
Version: master branch, rev 46dd028a3307
OS: macOS v10.13.6 (High Sierra)
QT: 5.12.3 from homebrew

When the drone gets killed by SIGTERM or SIGINT the cleanup is normally terminated with a segmentation fault, sometimes an illegal instruction error is shown.

I tried with a lot of combinations (as normal user, as root, with and without debugger, with and without interface filters, started from Ostinato or in a separate terminal window, using it with ostinato or just start then stop it). It always crashes at the end.

The operation before is flawless, I can transmit data without any issues.

The same tests work fine with v0.9 on this computer.

Here a log, when started with lldb:

```
/Users/behlers# lldb /Applications/Ostinato/drone.app/Contents/MacOS/drone
(lldb) target create "/Applications/Ostinato/drone.app/Contents/MacOS/drone"
Current executable set to '/Applications/Ostinato/drone.app/Contents/MacOS/drone' (x86_64).
(lldb) b main
Breakpoint 1: 22 locations.
(lldb) run
Process 12599 launched: '/Applications/Ostinato/drone.app/Contents/MacOS/drone' (x86_64)
Process 12599 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x00000001000196c9 drone`main(argc=1, argv=0x00007ffeefbff930) at drone_main.cpp:53
   50
   51  	int main(int argc, char *argv[])
   52  	{
-> 53  	    int exitCode = 0;
   54  	    QCoreApplication app(argc, argv);
   55
   56  	    app.setApplicationName("Drone");
Target 0: (drone) stopped.
(lldb) process handle SIGTERM -p true -s false
NAME         PASS   STOP   NOTIFY
===========  =====  =====  ======
SIGTERM      true   false  true
(lldb) cont
Process 12599 resuming
2019-08-18 00:38:33.040255+0200 drone[12599:1007590] Version: 0.9
2019-08-18 00:38:33.040291+0200 drone[12599:1007590] Revision: 46dd028a3307
PCAP Lib: libpcap version 1.8.1 -- Apple version 79.20.1
Retrieving the device list from the local machine

0. en0
Error setting direction(1) en0: Setting direction to PCAP_D_OUT is not supported on BPF

txStats: collection start
cap file = /private/var/folders/f7/lpy3m3gx32g1g6yfbm88y2fr0000gn/T/Drone.GPmTen
adding dev to all ports list <en0>
Mac: c82a1421cfc8
setRateAccuracy: rate accuracy set to Low - usleep
1. fw0
fw0 () rejected by filter. Skipping!
1. p2p0
p2p0 () rejected by filter. Skipping!
1. bridge0
bridge0 () rejected by filter. Skipping!
1. utun0
utun0 () rejected by filter. Skipping!
1. en1
en1 () rejected by filter. Skipping!
1. en2
en2 () rejected by filter. Skipping!
1. lo0
lo0 () rejected by filter. Skipping!
1. gif0
gif0 () rejected by filter. Skipping!
1. stf0
stf0 () rejected by filter. Skipping!
1. EHC250
EHC250 () rejected by filter. Skipping!
1. EHC253
EHC253 () rejected by filter. Skipping!
sysctl mib returns reqd len = 2276

if: lo0(1, 1)
if: gif0(2, 2)
if: stf0(3, 3)
if: EHC250(4, 4)
if: EHC253(5, 5)
if: en0(6, 6)
en0: failed to set promisc; SIOCSIFFLAGS failed (Operation not permitted)
if: en1(7, 7)
if: en2(8, 8)
if: fw0(9, 9)
if: p2p0(10, 10)
if: bridge0(11, 11)
if: utun0(12, 12)
port count = 12

stats for 12 ports setup
Port en0: ifIndex 6
host(add):  mac=c82a1421cfc8
Updater: User-Agent: Mozilla/5.0 (macOS High Sierra (10.13)) Drone/0.9 (Qt/5.12.3)
ComboProtocol: protoNumber = 208, 0x101112e40 <--> 0x101112f10
ComboProtocol: protoNumber = 206, 0x101113740 <--> 0x101113800
ComboProtocol: protoNumber = 206, 0x101113d90 <--> 0x101113e50
ComboProtocol: protoNumber = 207, 0x101113ce0 <--> 0x101113f20
ComboProtocol: protoNumber = 305, 0x101114a60 <--> 0x101114b70
ComboProtocol: protoNumber = 304, 0x101114f10 <--> 0x101115020
ComboProtocol: protoNumber = 303, 0x101115390 <--> 0x1011154a0
ComboProtocol: protoNumber = 306, 0x101115790 <--> 0x1011158a0
The server is running on 0.0.0.0: 7878
Version: 0.9
Revision: 46dd028a3307
Updater: latest version = 0.9
Process 12599 stopped and restarted: thread 1 received signal: SIGTERM
In ~BsdPort
In ~PcapPort
RxStats stop requested but is not running!
Receive stop requested but is not running!
txStats: collection end
Process 12599 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x0000000100004a2e drone`DeviceManager::~DeviceManager(this=0x0000000101401bf0) at devicemanager.cpp:113
   110 	        delete devGrp;
   111
   112 	    foreach(Device *dev, hostDeviceList_)
-> 113 	        delete dev;
   114 	}
   115
   116 	int DeviceManager::deviceGroupCount()
Target 0: (drone) stopped.
(lldb)
```

**Comment by b-ehlers** [comment]
Update: The same happens, when I start a second drone on the same port. The second drone issues an error message, starts it's cleanup function and then crashes.

Here the last messages, when starting the drone in the debugger:

```
ComboProtocol: protoNumber = 208, 0x10140b370 <--> 0x10140b440
ComboProtocol: protoNumber = 206, 0x10140bca0 <--> 0x10140bd60
ComboProtocol: protoNumber = 206, 0x10140c2f0 <--> 0x10140c3b0
ComboProtocol: protoNumber = 207, 0x10140c240 <--> 0x10140c480
ComboProtocol: protoNumber = 305, 0x10140cfc0 <--> 0x10140d0d0
ComboProtocol: protoNumber = 304, 0x10140d470 <--> 0x10140d580
ComboProtocol: protoNumber = 303, 0x10140d8f0 <--> 0x10140da00
ComboProtocol: protoNumber = 306, 0x10140dcf0 <--> 0x10140de00
Unable to start the server on <0.0.0.0>: The bound address is already in use
In ~BsdPort
In ~PcapPort
RxStats stop requested but is not running!
Receive stop requested but is not running!
txStats: collection end
Process 12724 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x8)
    frame #0: 0x0000000100004a2e drone`DeviceManager::~DeviceManager(this=0x0000000101201350) at devicemanager.cpp:113
   110 	        delete devGrp;
   111
   112 	    foreach(Device *dev, hostDeviceList_)
-> 113 	        delete dev;
   114 	}
   115
   116 	int DeviceManager::deviceGroupCount()
Target 0: (drone) stopped.
(lldb)
```

**Comment by pstavirs** [comment]
@b-ehlers Thanks for reporting the issue along with triage and debug info. That info helped me quickly identify the culprit and fix it.

Fixed by 635ca18bce9c7d9c4ed46aac6cab4f4d62bb6fd2 - appreciate if you can verify and confirm. Thanks in advance!

**Comment by b-ehlers** [comment]
Confirmed - your change fixes the issue, no more segfaults.
